### PR TITLE
Improving OS-specific parts of documentation

### LINKED ISSUE #842: Idea: improve platform-specific parts of the docs
Just came across this sphinx plugin: https://sphinxcontrib-platformpicker.readthedocs.io/en/stable/usage.html - this looks nice and could clean up some platform-specific parts of our docs

### PR DESCRIPTION
**Pull request by sayandip18**
## Contributor Checklist:

* [x] I have added or updated my entry in [the creators.json file](https://github.com/LiberTEM/LiberTEM/blob/master/packaging/creators.json)
* [x] I have added [a changelog entry](https://github.com/LiberTEM/LiberTEM/tree/master/docs/source/changelog) for my contribution
* [x] I have added/updated documentation for all user-facing changes
* [ ] I have added/updated test cases
* [ ] I have included the [rebuilt production build of the client](https://libertem.github.io/LiberTEM/contributing.html?#building-the-client) (only if
changes were made to the GUI)

### CODE REVIEW

**Comment by sayandip18** [comment]
Travis build failed. Gives the following `Extension error` inspite of adding `sphinxcontrib.platformpicker` to `extensions` in `conf.py` file

`Could not import extension sphinxcontrib.platformpicker (exception: No module named 'sphinxcontrib.platformpicker')`

Could this be some issue with sphinxcontrib-platformpicker itself?

**Comment by uellue** [comment]
> Gives the following Extension error inspite of adding sphinxcontrib.platformpicker to extensions in conf.py file

The extension should also be added as a dependency in `tox.ini` to the `docs`, `docs-check` and `docs-deploy` environments. By the way, did you already check the result locally using `tox -e docs`? On my system that reproduces the error. Note: If you change the environment specification in `tox.ini`or `test_requirements.txt`, you have to add the `-r` option to the `tox` command once to rebuild the environment.

**Comment by uellue** [comment]
...and many thanks for the pull request! :-)

**Comment by sayandip18** [comment]
No, haven't checked that yet. Will do it soon.

**Comment by sayandip18** [comment]
> 
> 
> By the way, did you already check the result locally using `tox -e docs`? On my system that reproduces the error. 


My system produces this error 

`ModuleNotFoundError: No module named 'click'`
`ERROR: InvocationError for command 'C:\Users\SANDIP\LiberTEM\.tox\docs\Scripts\python.EXE' 'scripts\build-authors-contributors' (exited with code 1)`

**Comment by uellue** [comment]
> ModuleNotFoundError: No module named 'click'

Hm, interesting. `click` is a dependency of LiberTEM for a long time already, so it _should_ be available in the environment that `tox` creates. Did you try rebuilding the environment using the `-r` option of `tox`? How exactly did you run the command, i.e. which OS, base python version, version of `tox` etc?

**Comment by sk1p** [comment]
> it _should_ be available in the environment that `tox` creates

Indeed, I just checked this: LiberTEM is installed into the `docs` environment, including all dependencies. Maybe a previous attempt left the environment in an incomplete state? Recreating using `tox -e docs -r` as suggested above should fix it, or reveal the underlying error

**Comment by sayandip18** [comment]
I ran `tox -e docs -r` and it produced this error

`ERROR: Failed building wheel for hdbscan`
`Successfully built libertem`
`Failed to build hdbscan`
`ERROR: Could not build wheels for hdbscan which use PEP 517 and cannot be installed directly`

though I've installed hdbscan using `conda install -c conda-forge hdbscan`

Also
Python version: 3.7.6
tox version: 3.17.1
conda version: 4.8.3
OS: Windows 10

**Comment by sk1p** [comment]
> I ran `tox -e docs -r` and it produced this error
> 
> `ERROR: Failed building wheel for hdbscan`
> `Successfully built libertem`
> `Failed to build hdbscan`
> `ERROR: Could not build wheels for hdbscan which use PEP 517 and cannot be installed directly`
> 
> though I've installed hdbscan using `conda install -c conda-forge hdbscan`

Thanks for the update!

Installing something via conda doesn't influence the tox environment, though: if you have a look at `.tox/docs`, there is a complete separate Python virtualenv. Python packages installed outside of this environment, for example via conda, are not available inside of this environment. 

> Python version: 3.7.6
> tox version: 3.17.1
> conda version: 4.8.3
> OS: Windows 10

I think the way to go here is, sadly, to install a C/C++ compiler toolchain, like visual studio community edition. Maybe @uellue can give a bit more details, as he is also running Windows.

**Comment by uellue** [comment]
> I think the way to go here is, sadly, to install a C/C++ compiler toolchain, like visual studio community edition. Maybe @uellue can give a bit more details, as he is also running Windows.

Yes, that is unavoidable at some point for serious work with Python on Windows. One needs a recent version of Visual Studio Community edition. VS is vast and can be quite a burden to install (think double-digit GB of download and disk space). One should pay attention to only install the parts that one REALLY needs for building packages, i.e. work with the detailed selection of packages and not the "solutions bundles", I don't remember what they are called exactly. As far as I remember, one needs only the very basic tools -- C/C++ compiler for the current platform, a few basic headers for Windows, and I think some form of build tool. One SHOULDN'T install things that are marked as Python-related in VS since that pulls a new install of Anaconda and can mess up existing Python installs. One has to be careful since selecting a package may mark a ton of dependencies for installation, and one has to unmark all of them by hand to get rid of them in the selection. Best is to install as little as possible, try to build, and then try to guess and google what might be missing with the error messages you get.

TL;DR: It may be faster to wipe the HDD, install Debian, and run something "apt install build-essential python3-dev"instead of getting Windows 10 ready for dev work... Just using WSL may also be an easy way out.

**Comment by sayandip18** [comment]
The problem is with headings here. Sphinxcontrib-platformpicker doesn't accept headings in the instruction block. I tested in a [sample PR](https://github.com/sayandip18/LiberTEM/pull/2) in my master branch. All the tests for [this test commit](https://github.com/sayandip18/LiberTEM/pull/2/commits/568f4bd93baf0f20a8e8c246823f77b15e80fa05), which had no heading in the instruction block after`:title: Windows`, were successful. 

I think one possible solution is to create a issue in sphinxcontrib-platformpicker. Otherwise we have to let go of the headings altogether.

**Comment by uellue** [comment]
One could add a platform picker not for all headlines in one, but separately for each section so that they don't go across headlines. If I remember the demo correctly, they all switch together.

**Comment by sayandip18** [comment]
This is odd! All the tests for the penultimate commit were successful. So, far `TOXENV=docs-check` was the only job failing. In the last commit, doc-check was successful, but `TOXENV=py36` has failed. To be more specific, [`test_disk_template.py`](https://github.com/LiberTEM/LiberTEM/blob/master/tests/template/test_disk_template.py) is the test which has failed.

**Comment by sk1p** [comment]
> This is odd! All the tests for the penultimate commit were successful. So, far `TOXENV=docs-check` was the only job failing. In the last commit, doc-check was successful, but `TOXENV=py36` has failed. To be more specific, [`test_disk_template.py`](https://github.com/LiberTEM/LiberTEM/blob/master/tests/template/test_disk_template.py) is the test which has failed.
> 
> I have only added the changelog in the last commit.

Yeah, I don't think the test fails because of your changes... @twentyse7en did you see [this error](https://travis-ci.com/github/LiberTEM/LiberTEM/jobs/366332990) before? It looks like for some reason the local cluster is not running or not completely running, like it is missing workers processes?

**Comment by twentyse7en** [comment]
> did you see [this error](https://travis-ci.com/github/LiberTEM/LiberTEM/jobs/366332990) before?

No

**Comment by sk1p** [comment]
I restarted the CI job and it appears to work; the test appears to be a bit flaky, so don't worry too much about it.

Regarding the implementation of the platform pickers: if there is a section that has multiple alternatives, they should be displayed in a single horizontal platform picker, like this:

image

Instead of having a platform picker for each platform, which results in this vertical list:

image

For AppImage or other platform-specific solutions, there could be a short description that this installation method is not available for the chosen platform.

Also, there appears to be something broken below the "using conda" headline - there are only headers displayed, and below "Updating", there is some syntax problem which causes some raw rst to leak through:

image

Did you manage to get the docs building on your system?

**Comment by sayandip18** [comment]
image
 
This is the only issue I'm facing now. This is a little strange as all the indentations are identical for all the headings and only this one is rendering in such a way.

**Comment by sayandip18** [comment]
The issue is resolved now. Any further improvements that you want me to make?

**Comment by sk1p** [comment]
> The issue is resolved now. Any further improvements that you want me to make?

Thanks for taking care of this! There are a few things that can still be improved:

image

In the AppImage section, the whole OS chooser is indented, it should be aligned to the left like the others. On macOS and Windows, I would prefer if the text read "only available on Linux" instead of "not available", to make it a bit more clear that it is available, just not on this OS.

For "Creating an isolated Python environment", the content is the same for all operating systems, so the paragraph should just not contain an OS chooser. This is a general thing - the "using virtualenv" section is the same for all OSes, so it doesn't need the OS selector (yet). Same for sections "PyTorch", "Other extra packages".

The "Using conda" section also has the indentation problem mentioned above - the whole section is indented. Also, the content seems to be mixed up: when I select macOS, I get something about Windows, and the Windows tab shows a Linux-like prompt. The content of the macOS tab also doesn't tell the user how to actually create an environment, only how to enable it - that needs to be adapted a bit, I think.

In the "CuPy" section, there appears to be an indentation in the "form git checkout" part on macOS - here, macOS should be like the Linux part, without the extra indentation after the environment name.

I think that's all for now, there are a few things that can be improved on top of that, but those can be done in a separate pull request, I think. Thanks again!

**Comment by sayandip18** [comment]
```
C:\Users\SANDIP\LiberTEM\docs\source\install.rst:333: WARNING: Explicit markup ends without a blank line; unexpected unindent.
```

I am getting this even though the line:333 in `install.rst` file looks like this:

image

**Comment by sk1p** [comment]
> ```
> C:\Users\SANDIP\LiberTEM\docs\source\install.rst:333: WARNING: Explicit markup ends without a blank line; unexpected unindent.
> ```

Hmm, I'm not sure what's happening here, is it possible you are mixing tabs and spaces? If not, could you push your current version to your branch so I can have a look? Thanks!

**Comment by sayandip18** [comment]
> 
> Hmm, I'm not sure what's happening here, is it possible you are mixing tabs and spaces? 

I use the space key all the time. I did use the tab key a few times, but not in this block. 

> 
>  If not, could you push your current version to your branch so I can have a look? 

Sure!

**Comment by sk1p** [comment]
> I use the space key all the time. I did use the tab key a few times, but not in this block.

It's not only what keys you press, but also your editor configuration: in most cases, it makes sense to convert tab key presses, and indent/dedent editor commands, into space characters, such that the indentation is the same regardless of editor others are using. In this case, the indentation is using tabs, except in line 333:

image

(this is highlighting all tab characters in that file)

Could you configure your editor to use 4 spaces instead, and convert all tabs to spaces? Then these kinds of errors should not happen anymore. If this is hard to do in your editor, I would suggest to use vs code instead, and use the CTRL+SHIFT+P "Convert indentation to spaces" command, which will do this conversion for you easily.

Hope this helps!

**Comment by sayandip18** [comment]
image

I edited the file from Visual Studio and and when I pushed the commit I saw this. Should I delete the entire .vs folder?

**Comment by sk1p** [comment]
About the `.vs` folder, you can also add it to the `.gitignore` file, so it is never added accidentally again.

**Comment by sayandip18** [comment]
I guess this can be merged now.

**Comment by uellue** [comment]
Thank you for working on this! When testing your version, I got confused which parts of the documentation are platform specific and which apply to all platforms. Is there a way to mark the parts that are switched by the platform picker?

Regarding virtualenvs: They do work on Windows, too. However, there we recommend using Conda. That means this part should get a note discouraging using virtualenvs when the Windows option is selected.

Before the PR can be merged, you should also rebase it on the current master branch and resolve the conflicts manually.

**Comment by sayandip18** [comment]
> Is there a way to mark the parts that are switched by the platform picker?

I don't think there is any, other than this
image


> Regarding virtualenvs: They do work on Windows, too. However, there we recommend using Conda. That means this part should get a note discouraging using virtualenvs when the Windows option is selected.

I didn't apply platformpicker for all the blocks. I left out virtualenv because it is more or less same for all the three OS.
> 
> Before the PR can be merged, you should also rebase it on the current master branch and resolve the conflicts manually.
 
I tried pulling changes from the master then pushing them.

**Comment by twentyse7en** [comment]
> I tried pulling changes from the master then pushing them.

You may need to [rebase](https://stackoverflow.com/questions/7929369/how-to-rebase-local-branch-with-remote-master/18442755#18442755) to master before pushing.

**Comment by sayandip18** [comment]
The branch got messed up yet again.

**Comment by sayandip18** [comment]
Closing in favour of #882
